src/scipp/core/assignments.py

```diff
@@ -0,0 +1,99 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+
+from __future__ import annotations
+
+from typing import Dict, Optional, Union
+
+from ..typing import DataArray, Dataset
+
+
+def assign_coords(
+    self, coords: Optional[Dict] = None, **coords_kwargs
+) -> Union[DataArray, Dataset]:
+    """Update or insert coordinates to ``scipp.DataArray`` or ``scipp.Dataset``.
+
+    Parameters
+    ----------
+    coords :
+        Coornidates to be updated or inserted to the
+        ``scipp.DataArray`` or ``scipp.Dataset``.
+
+    **coords_kwargs :
+        Keyword arguments form of ``coords``.
+
+
+    Returns
+    -------
+    :
+        ``scipp.DataArray`` or ``scipp.Dataset`` with updated coordinates.
+
+    """
+
+    if coords is not None:
+        collected_coords = {**coords, **coords_kwargs}
+    else:
+        collected_coords = coords_kwargs
+
+    for coord_key, coord in collected_coords.items():
+        self.coords[coord_key] = coord
```

Comment:
We should not change the input. Must make a shallow copy first (`self = self.copy(deep=False)`)